### ISSUE TITLE
Feature request : Wrap a C pointer with a Chapel array

### ISSUE THREAD

**Issue by npadmana** [chapel-lang/chapel#14656]
As a user, I would like to be able to take in an externally allocated array as a C pointer and treat it like a Chapel array. Since C pointers will not have any size/dimension information, the user would be responsible for specifying the domain of the array. Optionally, the user may also specify the type of the array (if eg. the allocated array is a void C pointer).

The use cases I can imagine are simple rectangular domains with no strides etc., so it would be very useful if this functionality worked for that specific case.

This captures a Stack Overflow discussion : https://stackoverflow.com/questions/53821252/best-way-to-wrap-a-c-array-pointer-as-a-chapel-array

**Comment by npadmana** [comment]
@bradcray - finally captured this discussion.

**Comment by mppf** [comment]
Actually this is already possible with the standard library function `makeArrayFromPtr`. However, this function is not documented (I think it slipped through the cracks somehow) and it should be assumed to be unstable.

Anyway, here is a complete example:

``` chapel
extern {
  #include <stdio.h>

  static double CArr[10];
  static double* CArrPtr = CArr;
  static void initCArr(void) {
    for (int i = 0; i < 10; i++) {
      CArr[i] = i;
    }
  }
  static void printCArr(void) {
    for (int i = 0; i < 10; i++) {
      printf("CArr[%i] = %lf\n", i, CArr[i]);
    }
  }
}

proc printArr(A: []) {
  for i in A.domain {
    writef("A[%i] = %r\n", i, A[i]);
  }
}

// Start by printing the C array which should be initialized to 0.
printCArr();

var ptr: c_ptr(c_double) = CArrPtr;
// Create a Chapel array that refers to the same memory as the C array.
// The memory is assumed to be managed externally (i.e, there will be
// no attempt to free 'ptr' when 'A' goes out of scope).
//
// The domain size must match the allocated region or else you will
// get memory errors.
var A = makeArrayFromPtr(ptr, {10..<20});

// Set the C array's elements from C and then print the Chapel array
// (to demonstrate that they point to the same data)
initCArr();
printArr(A);

// Set the array's elements in Chapel and then print it from C
// (to demonstrate again that they point to the same data)
A = A.domain;
printCArr();
```

**Comment by bradcray** [comment]
@mppf: I tried using this yesterday to see whether I could propose a solution that used it, but didn't end up getting an array that aliased the original array.  It could definitely have been user error, but before I try again to provide an example, have you validated that it shares the C array data?

**Comment by mppf** [comment]
Yes the program output demonstrates that it is aliasing the array.

First we have the initial C array data
```
CArr[0] = 0.000000
CArr[1] = 0.000000
CArr[2] = 0.000000
CArr[3] = 0.000000
CArr[4] = 0.000000
CArr[5] = 0.000000
CArr[6] = 0.000000
CArr[7] = 0.000000
CArr[8] = 0.000000
CArr[9] = 0.000000
```

Then we update the array from C and print it out from Chapel and see the updates:
```
A[10] = 0
A[11] = 1
A[12] = 2
A[13] = 3
A[14] = 4
A[15] = 5
A[16] = 6
A[17] = 7
A[18] = 8
A[19] = 9
```

And finally we update it from Chapel and print it out in C and see the updates:
```
CArr[0] = 10.000000
CArr[1] = 11.000000
CArr[2] = 12.000000
CArr[3] = 13.000000
CArr[4] = 14.000000
CArr[5] = 15.000000
CArr[6] = 16.000000
CArr[7] = 17.000000
CArr[8] = 18.000000
CArr[9] = 19.000000
```

**Comment by bradcray** [comment]
Whoops, sorry, I was reading too fast after seeing your Slack ping Michael and was only looking at the top-level code for assignments to the arrays, not into the called procedures (reason #n in why I dislike 9am meetings).

I'm not sure what I did wrong yesterday (it was on ATO and I didn't save the link), but checking again today, am finding it works: [ATO](https://ato.pxeger.com/run?1=fVPBbtNAED1w81fMDVsktIkptLE4oEiIA0hI5FZV0WY9jld1dq3xboKp-BIuvZSP4ms6u-s2lgJIlmy_fW925s3Mr9-yFi029_cPzlbTyz8v3M6UrkH40i8_qw3cwdkZrGrVAT8CKnGLsIRGbUhQnwC4jv9XfYtdkfBvS0bCFu0HItF_Uz_wI5mdD-TpaQZ3zAEgtI40XC3kWmlbMPbTi_eCQK4F0SK-RJ_KdWncpsEJXGWeF_Dr8xt4D7PXsyMyi8j8iMwjkh-R3CPzsepNREaqi4iMVG89ko9V7yIyUl1GJD-14Kul_zkg162llUmfysqiF-xGwrZ_QkJve28cG2NKdsG1pbBYgjXBeVsjaDwAGWeVxiQ5tmMydJBzkkZXassRdGehMZzrOaPe7Y475M3mLjD6774Vgc25LmLKz33JRrKTWqNq14dTJu54eMK3pzE75UiTkANTfcGyRnnLRQkLB3y55-J1jaRCwVwply68vS1S0ycHf9DodIh_GoIHWVW90tsgXkKYKO8nYdWg9FGVjmdhByIhGY3GNMzG3y8SuoS9kjjdI3UCUtf5m2bTjehC4BK_M5Alg-r64lVjhph5Ac9Bw21ZEbdvWMKnZXwE)  [ATO](https://ato.pxeger.com/run?1=fVPBjtMwED1wy1eMxIFEtF3asLDbigOqhDiAhERvK1S5zqSx1rWjidMSyn4Jl73AR_E1jO3sNlIBKVKS5_dmPG9mfvySlahR39__bF05vvr95PvOFq1G-NgtP6gNHOHiAlaVaoAfAaW4RViCVhsS1CUAbcP_q67GZpHwb01WwhbdWyLRfVbf8B3ZnQ_k6WkGR-YAELqWDFzP5VoZt2Dszov3gkCuBdE8vkSXynVh243GEVxnnhfwmxdf4A1MJ9MTMo3I7ITMIpKfkNwjs6HqZUQGqsuIDFSvPJIPVa8jMlBdRSQ_t-CTo_85INe1o5VNH8rKohfsRsK2v0dCb3tnWzbGFuxCWxfCYQHOBuddhWDwAGRbpwwmyakdo76DfCdpTam2HME0DrT1vjDq3W64Q95s7gKj_-7bIrD5rvN45ce-ZAPZWa1RtevCKRN3PDzh29OYnXKkERy1nUye-pvcscLXLSuUt1ybcHDAZ3v2wFRIKtTNBbMDwrtcI-kuOfgDbdI-zXkInmdVdspsg3gJYbC8rYSlRumjKhPPwipEQjKYkHEYkb8nEqaAvZI43iM1AtK28Zmm441oQuACvzKQJb3q5vK5tn3MfAGPQUO2bBGXsN_Fh538Aw).  I was very surprised it didn't yesterday, so regret not working harder at it.

I suspect part of the reason this procedure isn't documented may relate to it being defined in an internal rather than standard module.  I think we should probably move it into a standard module to address that—maybe `CTypes`?

**Comment by bradcray** [comment]
Oh, in editing my SO answer, I think I determined what I was doing yesterday that didn't seem to work.  I was trying to declare the new array's type explicitly, as in the SO question:

```chapel
var myArray: [1..size] real = …
```

which seems to result in a non-alias when using the "pass an integral size" version [ATO](https://ato.pxeger.com/run?1=fVNNaxsxED30tr9ibtkFr5O1-5HY9BAMpYcWCvUtBCNrx1kRWVpmtXbV0F_SSy7tj-qv6Uhy4gU3hYVFT_Nm5r0Z_fwtG9Gifnz81btNefnnld7autcIn_3ik1rDA5yfw7JRHfAnYCPuERag1ZoE-Qyg7_i89C1284yPLVkJd-iuiYT_qr7jB7LbkCiE5wU8cAwAoevJwNVMrpRxc8Z-BPJOEMiVIJqln_C5XNW2X2scwVUR4iJ-c3EL76EaV0ekSsjkiEwSMj0i04BMhqzXCRmw3iRkwHobkOmQ9S4hA9ZlQqanFnxx9D8H5Kp1tLT5k6wiecFuZGz7RyQMtnvbszG2Zhf6thYOa3A2Ou8aBIN7INs7ZTDLjuMYHSbIPQVfO55FsJX95l5fntA8RnNXs9Tc8wSKAe1EVWJtfbydwU01HoeCt6xTaCZueW3iXaAxO-fMo9gTU4NU2aC8ZznCwR7PdizbNEgqSmWNLFoEY1sk7bN9uNAmP9Q7TcErrDZembtIXkDcpeAk4UajDFmVSXdx-1NANliKMm7FvwsJU8NOSSx3SJ2AvO9Cpapciy4mrvEbA0V2YKWdKsNSwXPGWKqYp0d3eHtPb_Av) and a segfault when using the "pass a domain" version [ATO](https://ato.pxeger.com/run?1=fZPPattAEMYPvekp5lYJLCey-yexySEYSg8tFOpbCGa9GllL5F0xWtlVTc59iF5ySR4qT9PZXSUWuC0YjL6db2bnN7O_n2QpaqweHh5bW6QXz29-bU3eVghfu8UXtYYDnJ3BslQN8E9AIe4QFlCpNQnqIoC24e9lV2Mzj_izJiNhg_aaSHTf1U_8RGbrErnwOIEDxwAQ2pY0XM7kSmk7Z-3emXeCQK4E0Sz8iS6Wq9y06wpHcJm4OK_fnN_CFWTj7KhkQZkclUlQpkdl6pTJ0PUuKAPX-6AMXB-cMh26PgZl4LoIyvQUwTdL_yMgV7WlpYlf2koCC6YRMfbPSOiwd6ZlMCZnCm2dC4s5WOPJ2xJB4x7ItFZpjKLjOEb9BPlO0uhCbTiDbixUhu96zqqj3fCEHGyeAqv_ntvcR_NdZ-HKr3NJBraTXoNr2_nTGdxk47EreMvdi4qNW14mf-Zs7I458wgOfdg9-x0FWaK8406FhT2-3TERXSIpT4HbZx7CMa-Rqi7au4NKx33R0xS83arolN548wL8mjnIhEWF0mVVOpz5hxECosG-pH5h_l5I6Bx2SmK6Q2oExG3jKmXpWjQ-cY4_WEii3hXWLXX7Bq8ZfalkHt5j_yxfnucf).

Let me know if you can reproduce.

**Comment by mppf** [comment]
With `var A:[{10..<20}] real = makeArrayFromPtr(ptr, {10..<20});` it looks to me like they still alias but I get a segfault at the end of the program. Similar with `var A:[0..<10] real = makeArrayFromPtr(ptr, 10);`. There must be some related cases here that aren't working yet.

I get them not to alias with `var A:[1..10] real = makeArrayFromPtr(ptr, 10);` but that seems to be more of a Chapel array issue than an issue with `makeArrayFromPtr`. (These arrays have different domains, so maybe it makes sense that a full copy occurs?).

**Comment by bradcray** [comment]
> that seems to be more of a Chapel array issue than an issue with makeArrayFromPtr. (These arrays have different domains, so maybe it makes sense that a full copy occurs?).

Is this a case of stealing vs. not stealing in an array initialization context, do you think?  In any case, it seems like matching domains should not be required, though it also wouldn't surprise me if the current implementation did rely on that to be safe/conservative.  Specifically, as with the array-swap optimization, if the arrays are of conforming size/shape/distribution (which is to say, "legal to assign"), it seems like we should just be able to steal the ddata pointer independently of the domain identities or values.

**Comment by mppf** [comment]
> Is this a case of stealing vs. not stealing in an array initialization context, do you think? 

Yes. I looked at how it's compiling & what's running. The code I added in PR #25896 is only checking for domains with the same value (vs. with equivalent shape):

https://github.com/chapel-lang/chapel/blob/e76423d4ec241257631fe5ebd93cd7f3c30c8e78/modules/internal/ChapelArray.chpl#L3087-L3112

That was just me being conservative to get something merge-able. If we go to extend it, I would expect that there are additional constraints for the same shape; in particular while it's legal to assign arrays with different distributions, I don't think we can expect to "move" (or swap, for that matter) between arrays of the same shape but with different distributions (say, different Block bounding boxes). It looks like the swap implementation handles this for Block in `canDoOptimizedSwap` & we'd either need to do some rearranging so that `doiBuildArrayMoving` can also return failure in some way -- or we'd want to make that function handle moving or copying/deallocating. (To be clear, Block doesn't have a `doiBuildArrayMoving` yet; my PR only implemented it for DefaultRectangular).

IMO it'd be nice to have some well-defined & user-facing functions to ask:
 * do two domains refer to the same thing? (identity)
 * do two domains have the same shape? (i.e. is assignment allowed)
 * do two domains have the same alignment? (i.e. fast swap is allowed)

If we had a function to ask the 3rd one, we could use that here.
